# TP DE DATA PREPROCESSING
## Inspection complète du jeu de données Heart Disease Cleveland

**Étudiant :** Joseph MUSENGA  
**Promotion:** Master 1 A.I
**Jeu de données :** Heart Disease Cleveland

### Objectif

Ce travail consiste à réaliser une inspection complète du jeu de données *Heart Disease Cleveland* afin d'en comprendre la structure, la nature des variables, la qualité générale et les principales caractéristiques descriptives avant les étapes ultérieures de prétraitement des données.

## 1. Chargement et aperçu du jeu de données

Cette première étape consiste à importer les bibliothèques nécessaires, charger le jeu de données et effectuer un premier aperçu de son contenu.

In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

In [44]:
# Source publique du jeu de données via mon repository github
url = "https://raw.githubusercontent.com/jomusenga/data_preprocessing_tp/main/data/Heart_disease_cleveland_new.csv"

# Importation des données
source_data = pd.read_csv(url)

# Aperçu des données
source_data.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,0,145,233,1,2,150,0,2.3,2,0,2,0
1,67,1,3,160,286,0,2,108,1,1.5,1,3,1,1
2,67,1,3,120,229,0,2,129,1,2.6,1,2,3,1
3,37,1,2,130,250,0,0,187,0,3.5,2,0,1,0
4,41,0,1,130,204,0,2,172,0,1.4,0,0,1,0


In [45]:
DF=source_data.copy()

## 2. Structure générale du jeu de données

Cette section vise à examiner les dimensions du jeu de données, les variables disponibles, leurs types informatiques et le nombre de valeurs renseignées.

In [46]:
# Verifications des dimensions de notre jeu de données:
print("Nombre d'observations :", DF.shape[0])
print("Nombre de variables :", DF.shape[1])

Nombre d'observations : 303
Nombre de variables : 14


In [47]:
# On verifie la structure général de notre jeu de données
DF.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


**Commentaire:** La méthode `info()` utilisée ci-dessus nous a permis d’examiner la structure générale de notre jeu de données DF, notamment le nombre d’observations et de variables, les types de données ainsi que la présence éventuelle de valeurs manquantes. Le jeu de données contient 303 observations et 14 variables. Aucune valeur manquante n’est détectée à ce stade, puisque chaque variable compte 303 valeurs non nulles. Toutes les variables sont encodées sous un format numérique, dont 13 variables de type int64 et une variable de type float64 (oldpeak). Toutefois, le type numérique utilisé par Pandas ne signifie pas nécessairement que toutes les variables sont quantitatives : certaines peuvent représenter des variables catégorielles codées numériquement. Une inspection de leurs modalités permettra de préciser leur nature statistique.

## 3. Vérification de la qualité générale des données

Nous vérifions à présent la présence éventuelle de valeurs manquantes et d'observations dupliquées afin d'obtenir un premier aperçu de la qualité du jeu de données.

In [48]:
# On Vérifie le nombre de valeurs manquantes par variable
DF.isna().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

**Commentaire :** Le résultat confirme qu'aucune valeur manquante n'est présente dans les 14 variables du jeu de données.

In [49]:
# On compte les observations dupliquées
print(DF.duplicated().sum())

0


**Commentaire :** La méthode `duplicated()` permet d'identifier les observations entièrement dupliquées dans le jeu de données. Le résultat obtenu est égal à 0, ce qui indique qu'aucune observation dupliquée n'a été détectée.

## 4. Inspection de la nature et des modalités des variables

Bien que toutes les variables soient encodées numériquement, leur type informatique ne permet pas à lui seul de déterminer leur nature statistique. Nous examinons donc le nombre de valeurs distinctes ainsi que les modalités des variables afin de distinguer notamment les variables quantitatives des variables catégorielles codées numériquement.

In [50]:
DF.nunique().sort_values()

sex           2
fbs           2
target        2
exang         2
slope         3
thal          3
restecg       3
cp            4
ca            4
oldpeak      40
age          41
trestbps     50
thalach      91
chol        152
dtype: int64

La méthode `nunique()` a permis d’identifier le nombre de valeurs distinctes présentes dans chaque variable. Les résultats montrent que certaines variables possèdent très peu de modalités : `sex`, `fbs`, `target` et `exang` n’en comptent que 2 ; `slope`, `thal` et `restecg` en comptent 3 ; tandis que `cp` et `ca` en comptent 4. Bien que ces variables soient encodées numériquement, leur faible nombre de modalités indique que plusieurs d’entre elles représentent des variables catégorielles codées sous forme numérique. À l’inverse, `oldpeak`, `age`, `trestbps`, `thalach` et `chol` présentent respectivement 40, 41, 50, 91 et 152 valeurs distinctes, ce qui est cohérent avec leur caractère quantitatif. Cette inspection permet ainsi de mieux distinguer la nature statistique des variables au-delà de leur simple type informatique (`int64` ou `float64`).

In [51]:
# Affichage des modalités des variables ayant peu de valeurs distinctes
for col in DF.columns:
    if DF[col].nunique() <= 10:
        modalites = [int(x) for x in sorted(DF[col].unique())]
        print(f"{col} : {modalites}")

sex : [0, 1]
cp : [0, 1, 2, 3]
fbs : [0, 1]
restecg : [0, 1, 2]
exang : [0, 1]
slope : [0, 1, 2]
ca : [0, 1, 2, 3]
thal : [1, 2, 3]
target : [0, 1]


**Commentaire** : L'inspection des modalités confirme que certaines variables, bien qu'elles soient enregistrées sous un type numérique (int64), représentent en réalité des caractéristiques catégorielles codées numériquement. Les variables `sex`, `fbs`, `exang` et `target` sont binaires, tandis que `cp`, `restecg`, `slope` et `thal` possèdent plusieurs modalités codées par des nombres entiers. La variable `ca` prend les valeurs 0 à 3 et correspond à un nombre discret. Cette distinction entre type informatique et nature statistique des variables sera prise en compte dans les analyses ultérieures. Les principales variables quantitatives sont `age`, `trestbps`, `chol`, `thalach` et `oldpeak`.

### 4.1 Classification des variables

À la suite de l'inspection des modalités, les principales variables quantitatives retenues sont `age`, `trestbps`, `chol`, `thalach` et `oldpeak`.

Les variables `sex`, `cp`, `fbs`, `restecg`, `exang`, `slope`, `thal` et `target` correspondent à des variables catégorielles encodées numériquement. La variable `ca`, qui représente un nombre discret, sera considérée séparément selon les besoins des analyses ultérieures.

## 5. Statistiques descriptives

Après avoir identifié la nature des variables, nous examinons les principales caractéristiques descriptives des variables quantitatives : moyenne, écart-type, minimum, quartiles, médiane et maximum.

In [52]:
# Statistiques descriptives des variables quantitatives:
var_quantitatives=['age', 'trestbps', 'chol', 'thalach', 'oldpeak']
DF[var_quantitatives].describe().T

,count,mean,std,min,25%,50%,75%,max
age,303.0,54.438944,9.038662,29.0,48.0,56.0,61.0,77.0
trestbps,303.0,131.689769,17.599748,94.0,120.0,130.0,140.0,200.0
chol,303.0,246.693069,51.776918,126.0,211.0,241.0,275.0,564.0
thalach,303.0,149.607261,22.875003,71.0,133.5,153.0,166.0,202.0
oldpeak,303.0,1.039604,1.161075,0.0,0.0,0.8,1.6,6.2


**Commentaire**: L’analyse descriptive des variables quantitatives montre que les 303 observations sont renseignées pour chacune des cinq variables, ce qui confirme l’absence de valeurs manquantes pour celles-ci. L’âge (age) moyen des individus est d’environ 54,44 ans, avec une médiane de 56 ans. Les âges varient de 29 à 77 ans, et 50 % des individus ont un âge compris entre 48 et 61 ans. La pression artérielle au repos (trestbps) présente une moyenne d’environ 131,69 mmHg et une médiane de 130 mmHg. Les valeurs s’étendent de 94 à 200 mmHg, avec 50 % des observations comprises entre 120 et 140 mmHg. La valeur maximale de 200 est relativement éloignée du troisième quartile et mérite donc une vérification lors de l’analyse des valeurs aberrantes. Le cholestérol (chol) présente une moyenne d’environ 246,69 mg/dL et une médiane de 241 mg/dL. Les valeurs varient de 126 à 564 mg/dL, tandis que la moitié centrale des observations se situe entre 211 et 275 mg/dL. Le maximum de 564 mg/dL est particulièrement éloigné du troisième quartile et pourrait correspondre à une valeur extrême. La fréquence cardiaque maximale (thalach) est en moyenne de 149,61 battements par minute, avec une médiane de 153. Les valeurs sont comprises entre 71 et 202, et 50 % des observations se situent entre 133,5 et 166. nfin, oldpeak présente une moyenne d’environ 1,04, une médiane de 0,8 et des valeurs comprises entre 0 et 6,2. Son troisième quartile est de 1,6, alors que son maximum atteint 6,2. Cette différence, associée à une moyenne supérieure à la médiane, suggère une distribution potentiellement asymétrique vers les valeurs élevées et la présence possible de valeurs extrêmes. Dans l’ensemble, ces statistiques permettent d’identifier certaines valeurs potentiellement atypiques, notamment pour trestbps, chol et oldpeak. Leur caractère réellement aberrant devra toutefois être vérifié à l’aide de méthodes appropriées avant toute décision de correction ou de suppression.

### 5.1 Distribution des variables catégorielles

Nous examinons les effectifs et les proportions associés à chaque modalité.

In [53]:
var_categorielles=['sex', 'cp', 'fbs', 'restecg','exang', 'slope', 'thal', 'target']

for var in var_categorielles:
    
    effectifs = DF[var].value_counts().sort_index()
    pourcentages = (DF[var].value_counts(normalize=True).sort_index().mul(100).round(2))
    resume=pd.DataFrame({'Effectifs':effectifs,'Pourcentage (%)':pourcentages})

    print(f"\nVariable : {var}")
    display(resume)



Variable : sex


,Effectifs,Pourcentage (%)
sex,,
0,97,32.01
1,206,67.99



Variable : cp


,Effectifs,Pourcentage (%)
cp,,
0,23,7.59
1,50,16.50
2,86,28.38
3,144,47.52



Variable : fbs


,Effectifs,Pourcentage (%)
fbs,,
0,258,85.15
1,45,14.85



Variable : restecg


,Effectifs,Pourcentage (%)
restecg,,
0,151,49.83
1,4,1.32
2,148,48.84



Variable : exang


,Effectifs,Pourcentage (%)
exang,,
0,204,67.33
1,99,32.67



Variable : slope


,Effectifs,Pourcentage (%)
slope,,
0,142,46.86
1,140,46.20
2,21,6.93



Variable : thal


,Effectifs,Pourcentage (%)
thal,,
1,168,55.45
2,18,5.94
3,117,38.61



Variable : target


,Effectifs,Pourcentage (%)
target,,
0,164,54.13
1,139,45.87


**Commentaire :**
L’analyse des effectifs et des pourcentages permet d’examiner la répartition des observations entre les différentes modalités des variables catégorielles. Pour la variable *"sex"*, la modalité 1 est majoritaire avec 206 observations (67,99 %) contre 97 (32,01 %) pour la modalité 0. La variable *"cp"* est principalement représentée par la modalité 3 avec 144 observations (47,52 %), suivie des modalités 2 (28,38 %), 1 (16,50 %) et 0 (7,59 %). Pour *"fbs"*, la modalité 0 est très largement majoritaire avec 258 observations (85,15 %). Concernant *"restecg"*, les modalités 0 (49,83 %) et 2 (48,84 %) sont presque équitablement représentées, tandis que la modalité 1 est très peu fréquente (1,32 %). Pour *"exang"*, la modalité 0 représente 67,33 % des observations contre 32,67 % pour la modalité 1. La variable *"slope"* présente une répartition presque identique entre les modalités 0 (46,86 %) et 1 (46,20 %), alors que la modalité 2 ne représente que 6,93 % des observations. Pour *"thal"*, la modalité 1 est majoritaire avec 55,45 %, suivie de la modalité 3 avec 38,61 %, tandis que la modalité 2 est minoritaire (5,94 %). Enfin, la variable cible target présente une répartition relativement équilibrée : 164 observations (54,13 %) pour la modalité 0 contre 139 observations (45,87 %) pour la modalité 1. Dans l’ensemble, certaines variables présentent donc des modalités fortement déséquilibrées, notamment *"fbs"*, tandis que la variable cible *"target"* reste relativement équilibrée.

### 6. Synthèse générale de l'inspection

L'inspection du jeu de données *Heart Disease Cleveland* a permis d'en examiner la structure, la qualité générale, la nature des variables ainsi que leurs principales caractéristiques descriptives.

Le jeu de données est constitué de **303 observations et 14 variables**. L'examen de sa structure avec la méthode `info()` montre que toutes les variables sont encodées numériquement : **13 variables sont de type `int64` et une variable (`oldpeak`) est de type `float64`**. Cependant, l'analyse des valeurs distinctes et des modalités a montré que le type informatique numérique ne correspond pas nécessairement à une nature statistique quantitative. Plusieurs variables sont en réalité catégorielles et ont simplement été codées par des valeurs numériques.

L'inspection de la qualité générale des données n'a révélé **aucune valeur manquante** parmi les 303 observations. De même, **aucune observation entièrement dupliquée** n'a été détectée. Le jeu de données présente donc, à ce stade de l'inspection, une bonne complétude.

L'analyse des modalités a permis de distinguer les principales variables quantitatives, notamment `age`, `trestbps`, `chol`, `thalach` et `oldpeak`, des variables catégorielles codées numériquement telles que `sex`, `cp`, `fbs`, `restecg`, `exang`, `slope`, `thal` et `target`. La variable `ca`, qui prend les valeurs de 0 à 3, correspond quant à elle à un dénombrement discret et devra être considérée en fonction des besoins des analyses ultérieures.

Sur le plan descriptif, les individus ont un **âge moyen de 54,44 ans**, avec des âges compris entre **29 et 77 ans**. La pression artérielle au repos (`trestbps`) est en moyenne de **131,69**, le cholestérol (`chol`) de **246,69**, et la fréquence cardiaque maximale (`thalach`) de **149,61**. La variable `oldpeak` présente une moyenne de **1,04**. Certaines valeurs maximales, notamment **200 pour `trestbps`, 564 pour `chol` et 6,2 pour `oldpeak`**, apparaissent relativement éloignées des valeurs centrales. Elles constituent des valeurs extrêmes potentielles qui pourront être examinées plus précisément lors des étapes ultérieures du prétraitement des données, sans conclure à ce stade qu'elles sont erronées ou aberrantes.

L'inspection des variables catégorielles met également en évidence des répartitions différentes selon les variables. Certaines modalités sont fortement majoritaires, notamment pour `fbs`, dont la modalité 0 représente **85,15 %** des observations. À l'inverse, la variable cible `target` présente une répartition relativement équilibrée, avec **54,13 %** des observations pour la modalité 0 et **45,87 %** pour la modalité 1.

En conclusion, cette première inspection permet d'obtenir une compréhension globale de la structure et du contenu du jeu de données avant toute opération de transformation. Les données sont complètes et ne présentent pas de doublons entièrement identiques. Elles contiennent à la fois des variables quantitatives et des variables catégorielles encodées numériquement. Certains éléments, notamment les valeurs extrêmes potentielles et les différences de fréquence entre certaines modalités, devront faire l'objet d'analyses plus approfondies au fur et à mesure de l'avancement du cours de *Data Preprocessing* et ce, selon les orientations du Professeur.